composer/callbacks/generate.py

```diff
@@ -0,0 +1,116 @@
+
+"""Periodically log generations to wandb from a set of prompts."""
+from typing import Any, List, Union, cast
+
+from torch.nn.parallel import DistributedDataParallel
+from torch.utils.data import DataLoader, TensorDataset
+
+from composer.callbacks.utils import create_interval_scheduler
+from composer.core import Callback, Event, State, Time, get_precision_context
+from composer.loggers import Logger
+from composer.models import HuggingFaceModel
+from composer.utils import dist
+from composer.utils.import_helpers import MissingConditionalImportError
+
+
+class Generate(Callback):
+    """Periodically log generations from a set of prompts.
+
+    Args:
+        prompts (List[str]): The list of prompts you would like to produce generations for
+        interval (Union[str, int, :class:`.Time`]): The interval describing how often checkpoints should be
+            saved. If an integer, it will be assumed to be in :attr:`.TimeUnit.EPOCH`.
+            Otherwise, the unit must be either :attr:`.TimeUnit.EPOCH`, :attr:`.TimeUnit.BATCH`,
+            :attr:`.TimeUnit.TOKEN`, or :attr:`.TimeUnit.SAMPLE`.
+        batch_size (int): Size of a prompt batch for generation.
+        kwargs: All kwargs well be passed along to the call to generate. This is for things like `do_sample`, `top_p`, etc
+    """
+
+    def __init__(self, prompts: List[str], interval: Union[str, int, Time], batch_size: int, **kwargs: Any):
+        self.prompts = prompts
+        self.generate_kwargs = kwargs
+        self.batch_size = batch_size
+        self.check_interval = create_interval_scheduler(interval, include_end_of_training=False)
+
+    def run_event(self, event: Event, state: State, logger: Logger) -> None:
+        if state.get_elapsed_duration() is not None and self.check_interval(state, event):
+            self.generate(state, logger)
+        else:
+            super().run_event(event, state, logger)
+
+    def generate(self, state: State, logger: Logger):
+        model = state.model
+        if isinstance(model, DistributedDataParallel):
+            # Why does DistributedDataParallel wrap the HuggingFaceModel and not the other way around?
+            # Can we make this assumption?
+            model = model.module
+        assert isinstance(
+            model, HuggingFaceModel
+        ), f'Expected HuggingFaceModel, but got {state.model.__class__}'  # TODO: Extend to support any models that have a generate method.
+
+        tokenizer = model.tokenizer
+
+        try:
+            from transformers import PreTrainedTokenizer, PreTrainedTokenizerFast
+        except ImportError as e:
+            raise MissingConditionalImportError(extra_deps_group='nlp',
+                                                conda_package='transformers',
+                                                conda_channel='conda-forge') from e
+        Tokenizer = Union[PreTrainedTokenizer, PreTrainedTokenizerFast]
+        tokenizer = cast(Tokenizer, tokenizer)
```

Comment:
I believe it is because it fails the pyright test otherwise. Originally I had `assert isinstance(tokenizer, Tokenizer)`, but pyright is unable to infer its type from that for some reason. 

(The pyright error was something along the lines of tokenizer being a NoneType, so the accesses to its attributes below will fail. )

Reply:
cool cool, if pyright needs its thats fine